dart/common/Composite.h

```diff
@@ -0,0 +1,212 @@
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_COMPOSITE_H_
+#define DART_COMMON_COMPOSITE_H_
+
+#include "dart/common/detail/CompositeData.h"
+
+namespace dart {
+namespace common {
+
+/// Composite is a base class that should be virtually inherited by any class
+/// that wants to be able to manage Aspects.
+///
+/// The base Composite class is completely agnostic to what kind of Aspects it
+/// is given. Aspects are stored in a std::map, so access to its Aspects happens
+/// on average in log(N) time. Most often, a class that accepts Aspects will have
+/// certain Aspect types that it will need to access frequently, and it would be
+/// beneficial to have constant-time access to those Aspect types. To get
+/// constant-time access to specific Aspect types, you can use the templated
+/// class SpecializedForAspect.
+class Composite
+{
+public:
+
+  using State = detail::CompositeState;
+  using Properties = detail::CompositeProperties;
+
+  using AspectMap = std::map< std::type_index, std::unique_ptr<Aspect> >;
+  using RequiredAspectSet = std::unordered_set<std::type_index>;
+
+  template <typename... Aspects>
+  using MakeState = detail::MakeCompositeState<Aspects...>;
+
+  template <typename... Aspects>
+  using MakeProperties = detail::MakeCompositeProperties<Aspects...>;
+
+  /// Virtual destructor
+  virtual ~Composite() = default;
+
+  /// Default constructor
+  Composite() = default;
+
+  /// It is currently unsafe to copy an Composite
+  // TODO(MXG): Consider making this safe by cloning Aspects into the new copy
+  Composite(const Composite&) = delete;
+
+  /// It is currently unsafe to move an Composite
+  Composite(Composite&&) = delete;
+
+  /// It is currently unsafe to copy an Composite
+  Composite& operator=(const Composite&) = delete;
+
+  /// It is currently unsafe to move an Composite
+  Composite& operator=(Composite&&) = delete;
+
+  /// Check if this Composite currently has a certain type of Aspect
+  template <class T>
+  bool has() const;
+
+  /// Get a certain type of Aspect from this Composite
+  template <class T>
+  T* get();
+
+  /// Get a certain type of Aspect from this Composite
+  template <class T>
+  const T* get() const;
+
+  /// Make a clone of the aspect and place the clone into this Composite. If
+  /// an Aspect of the same type already exists in this Composite, the
+  /// existing Aspect will be destroyed.
+  template <class T>
+  void set(const T* aspect);
+
+  /// Use move semantics to place aspect into this Composite. If an Aspect of
+  /// the same type already exists in this Composite, the existing Aspect will
+  /// be destroyed.
+  template <class T>
+  void set(std::unique_ptr<T>&& aspect);
+
+  /// Construct an Aspect inside of this Composite
+  template <class T, typename ...Args>
+  T* createAspect(Args&&... args);
+
+  /// Remove an Aspect from this Composite.
+  template <class T>
+  void eraseAspect();
```

Comment:
I wonder if this function name can be renamed to `removeAspect()` since we've used `remove-` for this kind of function.